lib/elixir/lib/enum.ex

```diff
@@ -721,45 +721,33 @@ defmodule Enum do
   @spec fetch(t, index) :: {:ok, element} | :error
   def fetch(enumerable, index)
 
-  def fetch(first..last, index) when is_integer(index) do
-    fetch_range(first, last, index)
-  end
-
-  def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:ok, count} when (count + index) < 0 ->
-        :error
-
-      {:ok, count} ->
-        fetch_enumerable(enumerable, count + index, Enumerable)
-
-      {:error, _} ->
-        enumerable
-        |> reverse()
-        |> fetch_list((-index) - 1)
+  def fetch(enumerable, index) when is_list(enumerable) and is_integer(index) do
+    if index < 0 do
+      enumerable |> :lists.reverse |> fetch_list((-index) - 1)
+    else
+      fetch_list(enumerable, index)
     end
   end
 
-  def fetch(enumerable, index) when is_list(enumerable) and is_integer(index) do
-    fetch_list(enumerable, index)
+  def fetch(first..last, index) when is_integer(index) do
+    fetch_range(first, last, index)
   end
 
   def fetch(enumerable, index) when is_integer(index) do
-    count_result =
-      case Enumerable.count(enumerable) do
-        {:ok, count} when (count - 1 - index) < 0 ->
-          :error
-        {:ok, _} ->
-          {:ok, Enumerable}
-        {:error, module} ->
-          {:ok, module}
-      end
+    case Enumerable.count(enumerable) do
+      {:error, module} when index < 0 ->
```

Comment:
Each of these patterns have special handling for negative `index`, I think we should have a separate function clause for it similarly as we had before.